### PR TITLE
Init project for koa2

### PR DESCRIPTION
**Pull request by bananaappletw**
There are still many problems of migration to `koa2`.
Some modules doesn't work without using ES6 import syntax, Ex: https://github.com/koajs/csrf.
The should.js version of this repo is very old.
But the current should.js is not stable enough to upgrade.
So, I only migrate examples which run without errors.

### CODE REVIEW

**Comment by hemanth** [comment]
Nice to see the 99 PR being a nice refactor, these being a decent amount of breaking changes would need sometime for a detailed review, more over we could create another branch for `v2` and link it in the docs till it looks good.

**Comment by bananaappletw** [comment]
@hemanth Actually, this is @coderhaoxin 's idea.
Since the current koajs/koa default branch is `koa2`.
The examples project should be `koa2`, too.
We could try to migrate examples step by step on master branch.
Then more people could help this repo.

**Comment by coderhaoxin** [comment]
@bananaappletw great work !

> Since the current koajs/koa default branch is koa2. The examples project should be koa2, too.

Agreed :)

**Comment by bananaappletw** [comment]
@hemanth @coderhaoxin Any update?

**Comment by hemanth** [comment]
I would give a 👍

File: 404/app.js [hunk 1]
```diff
@@ -1,29 +1,29 @@
...
-      this.type = 'text';
-      this.body = 'Page Not Found';
+app.use(async (ctx, next) => {
+  try {
+    await next();
+    const status = ctx.status || 404;
+    if (status === 404) ctx.throw(404);
+  } catch (err) {
+    if (err.status !== 404) return;
+    ctx.status = err.status || 500;
```
Comment (coderhaoxin) [inline 404/app.js]: err.status must be `404` here. no need `|| 500`

File: 404/app.js [hunk 2]
```diff
@@ -1,29 +1,29 @@
...
-      break;
-    case 'json':
-      this.body = {
-        message: 'Page Not Found'
-      };
-      break;
-    default:
-      this.type = 'text';
-      this.body = 'Page Not Found';
+app.use(async (ctx, next) => {
```
Comment (coderhaoxin) [inline 404/app.js]: I prefer to keep function name `pageNotFound`, it's helpful for debug

File: 404/test.js [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
 
-var app = require('./app');
-var request = require('supertest').agent(app.listen());
-
-describe('404', function() {
-  describe('when GET /', function() {
-    it('should return the 404 page', function(done) {
+describe('404', () => {
+  describe('when GET /', () => {
+    it('should return the 404 page', done => {
```
Comment (coderhaoxin) [inline 404/test.js]: `suppertest` supports `promise` now.

File: Readme.md [hunk 4]
```diff
@@ -1,64 +1,40 @@
-# Koa Examples
+# Koa2 Examples
```
Comment (coderhaoxin) [inline Readme.md]: no need change `Koa` to `Koa2`

**Comment by coderhaoxin** [comment]
@bananaappletw great works! But can you finish most migrations in this PR ?

I don't think it's a good idea to remove most examples, and add them back later >_<.

Maybe we can also just keep koa@1 examples in a short time. cc @hemanth advices ?

**Comment by hemanth** [comment]
> But can you finish most migrations in this PR ?

Yes, that was the initial thought.

**Comment by bananaappletw** [comment]
@coderhaoxin @hemanth Thanks for review.
I will do the most migration, if they could be migrated.
Because there some examples which doesn't have relative package of `koa2` to implement same functionality of previous examples.
Maybe combine `koa1` and `koa2` examples together until `koa1` could be migrate to `koa2`?
Note in `README.md` which examples are not migrated yet?
What do you think?

**Comment by bananaappletw** [comment]
@coderhaoxin @hemanth I need some time to migrate it@@

File: 404/app.js (hunk 1 above)
Comment (bananaappletw) [inline 404/app.js]: I will change to `ctx.status = err.status;`

File: 404/test.js (hunk 3 above)
Comment (bananaappletw) [inline 404/test.js]: It's time to refactor the old test code.

**Comment by bananaappletw** [comment]
@coderhaoxin @hemanth I suggest just remove koa1 examples and note on `README.MD` those examples are not migrated yet.
Then this repo doesn't need to handle for different syntax (generator).
Other people could also help to migrate those examples.

**Comment by bananaappletw** [comment]
@coderhaoxin @hemanth Any feedback?
I think this is the compromise way for current situation.

**Comment by coderhaoxin** [comment]
> I suggest just remove koa1 examples and note on README.MD those examples are not migrated yet.

+1 👍

**Pull request closed without merging**